pallets/creator-staking/src/lib.rs

```diff
@@ -0,0 +1,748 @@
+    #[pallet::storage]
+    #[pallet::whitelist_storage]
+    #[pallet::getter(fn current_era)]
+    pub type CurrentEra<T> = StorageValue<_, EraIndex, ValueQuery>;
+
+    #[pallet::type_value]
+    pub fn ForceEraOnEmpty() -> Forcing {
+        Forcing::NotForcing
+    }
+
+    /// Mode of era forcing.
+    #[pallet::storage]
+    #[pallet::whitelist_storage]
+    #[pallet::getter(fn force_era)]
+    pub type ForceEra<T> = StorageValue<_, Forcing, ValueQuery, ForceEraOnEmpty>;
+
+    /// Stores the block number of when the next era starts
+    #[pallet::storage]
+    #[pallet::whitelist_storage]
+    #[pallet::getter(fn next_era_starting_block)]
+    pub type NextEraStartingBlock<T: Config> = StorageValue<_, T::BlockNumber, ValueQuery>;
+
+    #[pallet::storage]
+    #[pallet::getter(fn registered_creator)]
+    pub(crate) type RegisteredCreators<T: Config> =
+        StorageMap<_, Twox64Concat, SpaceId, CreatorInfo<T::AccountId>>;
+
+    /// Staking information about creator in a particular era.
+    #[pallet::storage]
+    #[pallet::getter(fn creator_stake_info)]
+    pub type CreatorEraStake<T: Config> = StorageDoubleMap<
+        _,
+        Blake2_128Concat,
+        SpaceId,
+        Twox64Concat,
+        EraIndex,
+        CreatorStakeInfo<BalanceOf<T>>,
+    >;
+
+    /// Info about stakers stakes on particular creators.
+    #[pallet::storage]
+    #[pallet::getter(fn staker_info)]
+    pub type GeneralStakerInfo<T: Config> = StorageDoubleMap<
+        _,
+        Blake2_128Concat,
+        T::AccountId,
+        Blake2_128Concat,
+        SpaceId,
+        StakerInfo<BalanceOf<T>>,
+        ValueQuery,
+    >;
+
+    /// General information about an era like TVL, total staked value, rewards.
+    #[pallet::storage]
+    #[pallet::getter(fn general_era_info)]
+    pub type GeneralEraInfo<T: Config> =
+        StorageMap<_, Twox64Concat, EraIndex, EraInfo<BalanceOf<T>>>;
+
+    /// General information about the staker.
+    #[pallet::storage]
+    #[pallet::getter(fn ledger)]
+    pub type Ledger<T: Config> =
+        StorageMap<_, Blake2_128Concat, T::AccountId, AccountLedger<BalanceOf<T>>, ValueQuery>;
+
+    /// Accumulator for block rewards during an era. It is reset at every new era
+    #[pallet::storage]
+    #[pallet::getter(fn block_reward_accumulator)]
+    pub type BlockRewardAccumulator<T> = StorageValue<_, RewardInfo<BalanceOf<T>>, ValueQuery>;
+
+    #[pallet::type_value]
+    pub fn RewardConfigOnEmpty() -> RewardDistributionConfig {
+        RewardDistributionConfig::default()
+    }
+
+    /// An active list of configuration parameters used to calculate reward distribution portions.
+    #[pallet::storage]
+    #[pallet::getter(fn reward_config)]
+    pub type ActiveRewardDistributionConfig<T> =
+        StorageValue<_, RewardDistributionConfig, ValueQuery, RewardConfigOnEmpty>;
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub (super) fn deposit_event)]
+    pub enum Event<T: Config> {
```

Comment:
Need to rethink the usage of the first parameter `who` on many events. I thought that we put the `origin` of the tx in the Event.who. But looks like it is used differently in different cases. This is confusing. Better consult with Max.